File: typescript/packages/management/src/tools/list_agent_tools.ts
import type { ToolContext, ToolHandler } from '../registrar.js';

/**
 * Surfaces the agent tool catalog grouped by app_slug.
 *
 * The backend returns the flat ``{tools, count}`` shape; this handler
 * adds the ``by_app`` map the spec promises (``{[app_slug]: Tool[]}``)
 * so callers can render per-provider sections without re-grouping
 * client-side. Optional ``app_slug`` input filters the flat list AND
 * the by_app map to a single provider.
 */
export const listAgentTools: ToolHandler = async (
  ctx: ToolContext,
  input: Record<string, unknown>,
) => {
  const filterAppSlug =
    typeof input.app_slug === 'string' ? input.app_slug : undefined;

  const catalog = await ctx.client.agents.listTools(ctx.sandboxId);
  const tools = filterAppSlug
    ? catalog.tools.filter((t) => t.app_slug === filterAppSlug)
    : catalog.tools;

  const by_app: Record<
    string,
    Array<{ name: string; description?: string | null }>
  > = {};
  for (const tool of tools) {
    const slug = tool.app_slug || 'unknown';
    if (!by_app[slug]) by_app[slug] = [];
    by_app[slug].push({ name: tool.name, description: tool.description ?? null });
  }

  return { tools, by_app, count: tools.length };
};

File: python/copass-management/src/copass_management/tools/list_agent_tools.py
from __future__ import annotations

from typing import Any, Dict, List, TYPE_CHECKING

if TYPE_CHECKING:
    from copass_management.registrar import ToolContext


async def list_agent_tools(ctx: "ToolContext", input: Dict[str, Any]) -> Any:
    """Surface the agent tool catalog grouped by ``app_slug``.

    Backend returns the flat ``{tools, count}`` shape; we add the
    ``by_app`` map the spec promises (``{app_slug: [Tool, ...]}``) so
    callers can render per-provider sections without re-grouping
    client-side. Optional ``app_slug`` input filters the flat list AND
    the ``by_app`` map to a single provider.
    """
    filter_slug = input.get("app_slug") if isinstance(input.get("app_slug"), str) else None

    catalog = await ctx.client.agents.list_tools(ctx.sandbox_id)
    raw_tools: List[Dict[str, Any]] = list(catalog.get("tools", []) if isinstance(catalog, dict) else [])
    if filter_slug:
        raw_tools = [t for t in raw_tools if t.get("app_slug") == filter_slug]

    by_app: Dict[str, List[Dict[str, Any]]] = {}
    for tool in raw_tools:
        slug = tool.get("app_slug") or "unknown"
        by_app.setdefault(slug, []).append(
            {"name": tool.get("name"), "description": tool.get("description")}
        )

    return {"tools": raw_tools, "by_app": by_app, "count": len(raw_tools)}
